server/server.go

```diff
@@ -0,0 +1,100 @@
+package server
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/vault/api"
+)
+
+// New returns an unstarted HTTP server with health and proxy handlers.
+func New(logger *log.Logger, vaultClient *api.Client) *http.Server {
+	mux := http.ServeMux{}
+	mux.HandleFunc("/_health", healthHandler)
+	mux.HandleFunc("/", proxyHandler(logger, vaultClient))
+	srv := http.Server{
+		Addr:    "127.0.0.1:8200",
+		Handler: &mux,
+	}
+
+	return &srv
+}
+
+// The proxyHandler is based on the Send function from Vault Agent's proxy:
+// https://github.com/hashicorp/vault/blob/22b486b651b8956d32fb24e77cef4050df7094b6/command/agent/cache/api_proxy.go
+func proxyHandler(logger *log.Logger, vaultClient *api.Client) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		switch r.Method {
+		case "GET":
+			cl, err := vaultClient.Clone()
+			if err != nil {
+				http.Error(w, err.Error(), 500)
+				return
+			}
+			cl.SetToken(vaultClient.Token())
+
+			body, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				http.Error(w, err.Error(), 500)
+				return
+			}
+
+			logger.Printf("Proxying %s %s\n", r.Method, r.URL.Path)
+			resp, err := proxyRequest(cl, r, body)
+			if err != nil {
+				http.Error(w, err.Error(), 502)
+				return
+			}
+
+			err = resp.Write(w)
+			if err != nil {
+				http.Error(w, "failed to write response back to requester", 500)
+				return
+			}
+			logger.Printf("Successfully proxied %s %s\n", r.Method, r.URL.Path)
+		default:
+			http.Error(w, fmt.Sprintf("unsupported method for vault-lambda-extension: %s", r.Method), 400)
+			return
+		}
+	}
+}
+
+func proxyRequest(client *api.Client, r *http.Request, body []byte) (*api.Response, error) {
+	// http.Transport will transparently request gzip and decompress the response, but only if
+	// the client doesn't manually set the header. Removing any Accept-Encoding header allows the
+	// transparent compression to occur.
+	r.Header.Del("Accept-Encoding")
+	client.SetHeaders(r.Header)
+
+	fwReq := client.NewRequest(r.Method, r.URL.Path)
+	fwReq.BodyBytes = body
+
+	query := r.URL.Query()
+	if len(query) > 0 {
+		fwReq.Params = query
+	}
+
+	resp, err := client.RawRequestWithContext(r.Context(), fwReq)
+	if err != nil {
+		errString := fmt.Sprintf("Failed to proxy request: %s", err)
+		if resp != nil {
+			defer resp.Body.Close()
+			body, err := ioutil.ReadAll(resp.Body)
+			errString += fmt.Sprintf("\n\n(%d) %s", resp.StatusCode, string(body))
```

Comment:
hah thank you, yeah I knew this wasn't right but failed to improve it on my own. I think you're right that it should be using `multierror` or error wrapping instead, but as it turns out I just deleted all of this special handling in response to the above comment anyway. The errors we're handling here include things like 403 from Vault, so we actually want to pass those responses on without modification.

Reply:
On the format of the body content, we don't concern ourselves with that, we just read bytes from the request body, forward them on as bytes, and do the same on the way back. I think it will almost always happen to be JSON? But we copy headers from the request and the response too, so that's all handled outside our proxy server.